### PR TITLE
Use 5-argument `mul!` instead of `mulαβ!`

### LINKED ISSUE #95: Watch for boundary cases from fast=true in glmm fits
If one of the constrained elements of θ converges to zero in the `fast=true` stage when fitting a `glmm` then it should be given a nonzero starting value in the general optimization.

### LINKED ISSUE #220: NLopt error for Bernoulli models with random slopes when NOT using fast=true
Something strange is happening with `Bernoulli` models. Normal seems to run okay.

Also, unrelated to this issue, but the package cannot handle arrays with missing values out of the box. 

```julia
using DataFrames, MixedModels, GLM, RDatasets, RData, RCall, Missings, CSV

lexdec = rcopy(R"languageR::lexdec")

fm = fit(MixedModel, @formula(Correct ~ 1 + Class*NativeLanguage*Frequency + Trial + (1+Class + Frequency + Trial|Subject) + (1|Word)), lexdec, Bernoulli())
ERROR: ArgumentError: invalid NLopt arguments: zero step size
Stacktrace:

julia> fm = fit!(GeneralizedLinearMixedModel(@formula(Correct ~ 1 + Class*NativeLanguage*Frequency + (1+Class + Frequency + Trial|Subject) + (1|Word)), lexdec, Bernoulli()))
ERROR: ArgumentError: invalid NLopt arguments: zero step size
Stacktrace:

julia> fm = fit!(GeneralizedLinearMixedModel(@formula(Correct ~ 1 + Class*NativeLanguage*Frequency + (1+Class + Frequency + Trial|Subject) + (1|Word)), lexdec, Bernoulli()), fast=true)
Generalized Linear Mixed Model fit by maximum likelihood (nAGQ = 1)
  Correct ~ 1 + Class + NativeLanguage + Frequency + Class & NativeLanguage + Class & Frequency + NativeLanguage & Frequency + Class & NativeLanguage & Frequency + (1 + Class + Frequency + Trial | Subject) + (1 | Word)
  Distribution: Bernoulli{Float64}
  Link: LogitLink()

  Deviance: 475.6604

Variance components:
             Column        Variance      Std.Dev.    Corr.
 Subject (Intercept)   1.8648165158988 1.365582848
         Class: plant  0.6393674729270 0.799604573 -0.49
         Frequency     0.1061836740609 0.325858365 -0.94  0.73
         Trial         0.0000076274775 0.002761789  0.93 -0.21 -0.75
 Word    (Intercept)   0.7313290831756 0.855177808
 Number of obs: 1659; levels of grouping factors: 21, 79

Fixed-effects parameters:
──────────────────────────────────────────────────────────────────────────────────────────
                                                   Estimate  Std.Error    z value  P(>|z|)
──────────────────────────────────────────────────────────────────────────────────────────
(Intercept)                                       -1.11513    0.839064  -1.32902    0.1838
Class: plant                                      -0.721878   1.14746   -0.629111   0.5293
NativeLanguage: Other                              0.913713   1.03829    0.88002    0.3788
Frequency                                         -0.495801   0.174133  -2.84725    0.0044
Class: plant & NativeLanguage: Other               1.20139    1.33366    0.900816   0.3677
Class: plant & Frequency                          -0.140558   0.274048  -0.512895   0.6080
NativeLanguage: Other & Frequency                 -0.118276   0.221975  -0.532835   0.5941
Class: plant & NativeLanguage: Other & Frequency  -0.263183   0.336882  -0.781232   0.4347
──────────────────────────────────────────────────────────────────────────────────────────


```

### PR DESCRIPTION
**Pull request by dmbates**
Starting in Julia v1.3.0 methods for `LinearAlgebra.mul!` can be defined for 5 arguments, the last two being scalar multiples of `A*B` and of `C`.  These are used in the blocked Cholesky factorization in the `updateL!` method for this package, which is the computational workhorse.  Previous versions of this package defined a generic called `mulαβ!` for this but now that functionality can be expressed as methods for `mul!`.

* Restrict to Julia version 1.3 and later
* Reformat source code using the `Formatter` package
* Drop badge for coveralls.io in README page as coveralls is not configured for this repository
* GLMMs with `fast=false` (the default) no longer use a 2-stage optimization (fixes #220, closes #201, closes #95)

### CODE REVIEW

File: src/generalizedlinearmixedmodel.jl [hunk 1]
```diff
@@ -211,51 +262,72 @@ end
...
+    wts = [],
+    offset = [],
+    contrasts = Dict{Symbol,Any}(),
+)
     if isa(d, Binomial) && isempty(wts)
         d = Bernoulli()
     end
     (isa(d, Normal) && isa(l, IdentityLink)) &&
-        throw(ArgumentError("use LinearMixedModel for Normal distribution with IdentityLink"))
+    throw(ArgumentError("use LinearMixedModel for Normal distribution with IdentityLink"))
```
Comment (palday) [inline src/generalizedlinearmixedmodel.jl]: see prev comments on indent

File: src/linalg/rankUpdate.jl [hunk 2]
```diff
@@ -66,32 +84,40 @@ function rankUpdate!(C::Diagonal{T}, A::SparseMatrixCSC{T}, α=true, β=true) wh
...
     Ac = A.cscmat
     cp = Ac.colptr
-    all(diff(cp) .== S) ||
-        throw(ArgumentError("Each column of A must contain exactly S nonzeros"))
-    j,k,l = size(C.data.data)
+    all(diff(cp) .== S) || throw(ArgumentError("Each column of A must contain exactly S nonzeros"))
+    j, k, l = size(C.data.data)
     S == j == k && div(Ac.m, S) == l ||
-        throw(DimensionMismatch("div(A.cscmat.m, S) ≠ length(C.data.facevec)"))
+    throw(DimensionMismatch("div(A.cscmat.m, S) ≠ length(C.data.facevec)"))
```
Comment (palday) [inline src/linalg/rankUpdate.jl]: indent?

File: src/linearmixedmodel.jl [hunk 3]
```diff
@@ -420,15 +494,40 @@ Return negative twice the log-likelihood of model `m`
 """
 function objective(m::LinearMixedModel)
     wts = m.sqrtwts
-    logdet(m) + dof_residual(m)*(1 + log2π + log(varest(m))) - (isempty(wts) ? 0 : 2sum(log, wts))
+    logdet(m) + dof_residual(m) * (1 + log2π + log(varest(m))) -
+    (isempty(wts) ? 0 : 2 * sum(log, wts))
```
Comment (palday) [inline src/linearmixedmodel.jl]: indent

File: src/linearmixedmodel.jl [hunk 4]
```diff
@@ -471,15 +580,15 @@ end
 ranef!(v::Vector, m::LinearMixedModel, uscale::Bool) = ranef!(v, m, fixef(m), uscale)
 
 """
-    ranef(m::LinearMixedModel; uscale=false) #, named=true)
+    ranef(m::LinearMixedModel; uscale=false, named=true)
```
Comment (palday) [inline src/linearmixedmodel.jl]: this signature doesn't match the function below for `named`

File: src/optsummary.jl [hunk 5]
```diff
@@ -38,14 +38,36 @@ mutable struct OptSummary{T <: AbstractFloat}
     feval::Int
     optimizer::Symbol
     returnvalue::Symbol
-    nAGQ::Integer           # doesn't really belong here but I needed some place to store it
+    nAGQ::Integer           # don't really belong here but I needed a place to store them
```
Comment (palday) [inline src/optsummary.jl]: I would leave the singular verb and pronoun here.

File: src/remat.jl [hunk 6]
```diff
@@ -232,12 +232,12 @@ end
...
     mul!(Matrix{T}(undef, rank(adjA.parent), size(B, 2)), adjA, B)
-
+#=
 LinearAlgebra.mul!(C::AbstractMatrix{T}, adjA::Adjoint{T,<:FeMat{T}},
-        B::ReMat{T}) where {T} = mulαβ!(C, adjA, B)
-
-function mulαβ!(C::Matrix{T}, adjA::Adjoint{T,<:FeMat{T}}, B::ReMat{T,1},
-        α=true, β=false) where {T}
+        B::ReMat{T}) where {T} = mul!(C, adjA, B)
+=#
```
Comment (palday) [inline src/remat.jl]: do we need to preserve the code commented out here?

File: src/varcorr.jl [hunk 7]
```diff
@@ -1,11 +1,11 @@
 """
-VarCorr
+    VarCorr
 
 Information from the fitted random-effects variance-covariance matrices.
 
 # Members
 * `σρ`: a `NamedTuple` of `NamedTuple`s as returned from `σρs`
-* `s`: the estimate of the scale parameter in the distribution of the conditional dist'n of Y
+* `s`: the estimate of the per-observation dispersion parameter
```
Comment (palday) [inline src/varcorr.jl]: This is a slightly better formulation for GLMM, but I'm not sure which would be clearer to a normal user. Dispersion vs. scale is a toss-up for me; conditional distribution of Y is quite clear to me, but perhaps less to a normal user. What about "observation-level" instead of "per-observation"?

File: test/pirls.jl [hunk 8]
```diff
@@ -46,12 +39,12 @@ end
...
-    @test isapprox(sum(abs2, gm2.u[1]), 9.723175126731014, atol=0.0001)
-    @test isapprox(logdet(gm2), 16.90099, atol=0.0001)
-    @test isapprox(sum(gm2.resp.devresid), 73.47179193718736, atol=0.001)
-    @test isapprox(loglikelihood(gm2), -92.02628186555876, atol=0.001)
+    gm2 = fit(MixedModel, @formula(prop ~ 1 + p + (1|h)), cbpp, Binomial(), wts=cbpp[!,:s])
+    @test isapprox(deviance(gm2,true), 100.09585619892968, atol=0.0001)
+    @test isapprox(sum(abs2, gm2.u[1]), 9.723054788538546, atol=0.0001)
+    @test isapprox(logdet(gm2), 16.90105378801136, atol=0.0001)
+    @test isapprox(sum(gm2.resp.devresid), 73.47174762237978, atol=0.001)
+    @test isapprox(loglikelihood(gm2), -92.02628186840045, atol=0.001)
```
Comment (palday) [inline test/pirls.jl]: Why did the reference values change?

File: test/pirls.jl [hunk 9]
```diff
@@ -63,15 +56,14 @@ end
     @test lowerbd(gm3) == vcat(fill(-Inf, 6), zeros(2))
     @test fitted(gm3) == predict(gm3)
     # these two values are not well defined at the optimum
-    @test sum(x -> sum(abs2, x), gm3.u) ≈ 273.31563469936697 rtol=1e-3
-    @test sum(gm3.resp.devresid) ≈ 7156.558983084621 rtol=1e-4
+    @test isapprox(sum(x -> sum(abs2, x), gm3.u), 273.29266717430795, rtol=1e-3)
+    @test sum(gm3.resp.devresid) ≈ 7156.547357801238 rtol=1e-4
```
Comment (palday) [inline test/pirls.jl]: ibid

**Comment by palday** [comment]
I think we're still missing tests for one-stage `fast=false`. I'll add them in.

File: src/arraytypes.jl [hunk 10]
```diff
@@ -45,7 +45,7 @@ function Base.getindex(A::UniformBlockDiagonal{T}, i::Int, j::Int) where {T}
     Ad = A.data
     m, n, l = size(Ad)
     (0 < i ≤ l * m && 0 < j ≤ l * n) ||
-        throw(IndexError("attempt to access $(l*m) × $(l*n) array at index [$i, $j]"))
+    throw(IndexError("attempt to access $(l*m) × $(l*n) array at index [$i, $j]"))
     iblk, ioffset = divrem(i - 1, m)
```
[the remark this answers is not in the document]
Comment (dmbates) [reply src/arraytypes.jl]: I don't have strong opinions on that.  I am going with what the `Formatter` package does just to make it easier to maintain.  In Atom/Juno one can use `<Ctrl>-j <Ctrl>-f` to format a file and I kind of like the simplicity of that.

File: src/remat.jl (hunk 6 above)
Comment (dmbates) [reply src/remat.jl]: No, I should have removed it

File: src/optsummary.jl (hunk 5 above)
Comment (dmbates) [reply src/optsummary.jl]: I meant to indicate that the two last values in the structure are not related to the optimization per se but are tucked away in here for lack of a better place to put them.

File: src/varcorr.jl (hunk 7 above)
Comment (dmbates) [reply src/varcorr.jl]: "observation-level" would be fine.  I used "dispersion" because that is the name used in the `GLM` package.

File: test/pirls.jl (hunk 8 above)
Comment (dmbates) [reply test/pirls.jl]: Slightly different paths for the optimization because of different starting estimates.  It seems that the deviance at the optimum is reasonably well determined but the individual components can trade off against one another to some extent.

Comment (palday) [reply test/pirls.jl]: Right, that might explain the difficulties I'm having with the tests I'm working on -- the lme4 estimates have deviance close to those from this branch, but the estimates vary a fair amount. And it might also explain why the old gamma tests in #201 gave different answers on different platforms.

**Comment by palday** [comment]
Something that would be problematic under the old two-stage method perhaps (I have some things from `lexdec`), i.e. the usual "show something didn't work but now does" test, although I'm not sure how much use that is here. The `lexdec` dataset will be part of the tests for gamma regression, which I've been tinkering with.

File: src/varcorr.jl (hunk 7 above)
Comment (palday) [reply src/varcorr.jl]: Then let's do "observation-level dispersion".

Comment (palday) [reply src/varcorr.jl]: actually, having just searched through the project ... "per-observation" is already used extensively in the docs, so let's just leave it alone.

**Comment by palday** [comment]
I've got the models for such a test sketched out in my [fork](https://github.com/JuliaStats/MixedModels.jl/compare/dropmulalphabeta...palday:dropmulalphabeta) and can move them over if you think testing the old boundary cases makes sense. Right now, MixedModels.jl and lme4 deliver somewhat different answers for the Bernoulli model and the correct answer for the Gamma model remains elusive.
